torch/testing/_internal/mps_utils.py

```diff
@@ -0,0 +1,988 @@
+        except TypeError as e:
+            raise unittest.SkipTest("Uses dtype not supported on MPS") from e
+        except unittest.SkipTest as e:
+            # Don't error out on tests that have been explicitly skipped for some other reason
+            raise e
+        except Exception as e:
+            warnings.warn(
+                "Test is marked as unimplemented on MPS,"
+                "but instead of NotImplementedError or TypeError we received {type(e).__name__}:{e} "
+            )
+            raise unittest.SkipTest(
+                "[WARNING]: Test is marked as unimplemented on MPS,"
+                "but instead of NotImplementedError or TypeError we received {type(e).__name__}:{e} "
+            ) from e
+        else:
+            raise RuntimeError(
+                "Test is marked as unimplemented on MPS, but received unexpected success. Ensure CPU fallback is disabled"
+            )
+
+    return wrapper
+
+
+class MPSSkipInfo:
+    def __init__(
+        self,
+        *args: str,
+        test_class: Optional[str] = None,
+        variant: Optional[str] = None,
+        dtypes: Optional[Union[torch.dtype, list[torch.dtype]]] = None,
+        skip: Callable = unittest.expectedFailure,
+        skip_msg: str = "Skipped!",
+        upper: Optional[float] = None,
+        lower: Optional[float] = None,
+    ):
+        """Basic struct for tracking MPS OpInfo xfails
+
+        args: String names of test(s) to apply this xfail info to
+        test_class: Test class, e.g. 'TestCommon' etc.
+        variant: Variant name. Set to empty str ("") to explicitly specify the non-variant case
+        If set to None, will instead apply to all variants of the test
+        dtypes: If none specified, xfails all dtype variants
+        skip: Type of decorator to add [expectedFailure, skipTest, xfailUnimplementedOpMPS, xfailUnimplementedDtypeMPS]
+        upper: Upper bound MacOS version this xfail applies to (exclusive)
+        lower: Lower bound MacOS version this xfail applies to (inclusive)
+        """
+        self.tests: list[str] = []
+        for arg in args:
+            self.tests.append(arg)
+        self.test_class = test_class
+        self.variant = variant
+        self.dtypes = dtypes
+        self.skip = skip
+        self.skip_msg = skip_msg
+        self.upper = upper
+        self.lower = lower
+
+        if UNIMPLEMENTED in self.tests:
+            # Not all tests fail for unimplemented ops, so we specify here
+            self.tests = [TEST_OUTPUT_GRAD_MATCH, TEST_OUTPUT_MATCH]
+            self.skip = xfailUnimplemented
+
+
+""" Failures due to lack of op implementation on MPS backend """
+UNIMPLEMENTED_XFAILIST = {
+    "login": MPSSkipInfo(UNIMPLEMENTED),
+    "logspace": MPSSkipInfo(UNIMPLEMENTED),
+    "logspacetensor_overload": MPSSkipInfo(UNIMPLEMENTED),
+    "linalg.eig": MPSSkipInfo(UNIMPLEMENTED),
+    "linalg.eigvals": MPSSkipInfo(UNIMPLEMENTED),
+    "put": MPSSkipInfo(UNIMPLEMENTED),
```

Comment:
Please use generator expressions rather than repeating the same value over and over again
```suggestion
UNIMPLEMENTED_XFAILIST = { x :  MPSSkipInfo(UNIMPLEMENTED) for x in [
    "login", "logspace", "logspacetensor_overload", "linalg.eig", "linalg.eigvals", "put",] } + {
```

Reply:
A lot of the values in the list are repeated and could be codegen'd instead, but some of the list elements do have different values. I could codegen part of the list, and then specify the nonstandard values, but I thought that might be a bit confusing for anyone else who has to go in and update the skip list